rfcs/20251203-grouped-gemm-support/README.md

```diff
@@ -0,0 +1,604 @@
+
+## Frameworks Implementations
+
+### PyTorch Grouped Scaled MM
+
+PyTorch provides [`torch/nn/functional.py`](https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py) API for grouped scaled matrix multiplication.
+The `torch._scaled_grouped_mm` API is used internally:
+
+```python
+torch._scaled_mm_grouped(
+    mat_a,                           # Input tensor A (concatenated groups)
+    mat_b,                           # Input tensor B (concatenated groups)
+    scale_a,                         # Scaling factors for mat_a
+    scale_recipe_a,                  # Scaling type for mat_a, choises are: PER_TENSOR, PER_ROW, PER_COLUMN, AXISWISE
+    scale_b,                         # Scaling factors for mat_b
+    scale_recipe_b,                  # Scaling type for mat_b
+    swizzle_a=None,                  # Swizzle pattern for scale_a
+    swizzle_b=None,                  # Swizzle pattern for scale_b
+    bias=None,                       # Optional bias tensor
+    offs=None,                       # Device-side offsets for source tensor
+    output_dtype=torch.bfloat16,     # Output data type
+    contraction_dim=None,            # Which dims are K in matmul
+    use_fast_accum=False             # Nvidia GPUs specifics
+)
+```
+[See also TorchAO](https://github.com/pytorch/ao/tree/main/torchao/prototype/moe_training)
+for drop-in replacement with various quantization schemes.
+
+Memory layout:
+```
+mat_a:  [Expert0: M0xK | Expert1: M1xK | Expert2: M2xK]
+offs:    ^              ^              ^
+         0              M0             M0 + M1
+Weights: [num_experts x K x N]
+```
+
+### OpenVINO MoE
+
+OpenVINO implements a complete MoE operation in [PR #32469](https://github.com/openvinotoolkit/openvino/pull/32469):
+
+| Functionality | Description |
+|---------------|-------------|
+| `moe_mask_gen` | Generates per-expert token assignments and offsets from Top-K routing |
+| `moe_gather` | Gathers tokens by expert assignment for contiguous memory access |
+| `moe_gemm` | Performs expert-specific GEMMs with INT4/INT8/FP16 quantization |
+| `moe_scatter_reduction` | Scatters expert outputs back to original positions with weighted reduction |
+
+Memory layout uses per-expert indexing with concatenated buffers:
+
+```cpp
+float* input_data;              // Contiguous by active experts only
+int* expert_id;                 // Dynamic size, IDs of experts with tokens
+int* experts_info_start_idx;    // Start offset for each expert's tokens
+int* tokens_lens_per_expert;    // Number of tokens per expert
+float* weight_data;             // [Expert0_weights | Expert1_weights | ...]
+```
+
+- Offsets and sizes have dynamic size based on active ((!) not total) experts (therefore extra array is needed),
+  that is a bit different from the PyTorch/proposed oneDNN design, but compatible.
+- Current version provides offsets on host,
+  but device-side offsets expected for the future GPU-based routing.
```

Comment:
This PR (https://github.com/openvinotoolkit/openvino/pull/33073) makes `moe_mask_gen` to run on GPU, and offsets to be on usm_device memory.